### ISSUE TITLE
Java 9 Migration

### ISSUE THREAD

**Issue by ilovezfs** [Homebrew/homebrew-core#19696]
This will serve as a tracking issue for the Java 9 migration.

Broken at run time (`brew test` fails or other breakage observed):
(Check mark indicates it was either fixed or pinned to Java 8)
- [x] amdatu-bootstrap https://github.com/Homebrew/homebrew-core/pull/22609
- [x] ammonite-repl https://github.com/Homebrew/homebrew-core/pull/19699
- [ ] apache-flink https://gist.github.com/ilovezfs/d55534183d8fef0e7d93703bc84309db
- [x] apache-spark https://github.com/Homebrew/homebrew-core/pull/22194
- [x] app-engine-java https://github.com/Homebrew/homebrew-core/pull/20250
- [x] aspectj https://github.com/Homebrew/homebrew-core/pull/20774
- [x] bazel https://github.com/Homebrew/homebrew-core/pull/20544
- [x] druid https://github.com/Homebrew/homebrew-core/pull/21344
- [x] ecj https://github.com/Homebrew/homebrew-core/pull/28368
- [x] elasticsearch@2.4 https://gist.github.com/ilovezfs/c6d8a7900fa27922fcd55c27cbe69096 https://github.com/Homebrew/homebrew-core/pull/26876
- [x] elasticsearch@5.6 https://github.com/Homebrew/homebrew-core/pull/26876
- [x] embulk https://github.com/Homebrew/homebrew-core/pull/23881
- [x] flyway https://github.com/Homebrew/homebrew-core/pull/21601
- [x] javarepl https://github.com/Homebrew/homebrew-core/pull/24032
- [x] jing-trang https://github.com/Homebrew/homebrew-core/pull/24033
- [x] jmeter https://github.com/Homebrew/homebrew-core/pull/24034
- [x] jooby-bootstrap https://github.com/Homebrew/homebrew-core/pull/28311
- [x] jvm-mon https://github.com/Homebrew/homebrew-core/pull/24035
- [x] kobalt https://github.com/Homebrew/homebrew-core/pull/20441
- [x] liquigraph https://github.com/Homebrew/homebrew-core/pull/22227
- [x] logstash https://github.com/Homebrew/homebrew-core/pull/19270
- [x] maven https://gist.github.com/ilovezfs/976c6ae6419032695dec14508ab4a986
- [x] maven@3.0 https://gist.github.com/ilovezfs/f8435f90e2dbb328a2e5bae9df2ac803
- [x] maven@3.3 https://gist.github.com/ilovezfs/138feed3e3884c482f8c7b29fb6def3b
- [x] metabase https://github.com/Homebrew/homebrew-core/pull/21754
- [x] mvnvm https://github.com/Homebrew/homebrew-core/pull/20195
- [x] neo4j https://github.com/Homebrew/homebrew-core/pull/19682
- [x] payara https://github.com/Homebrew/homebrew-core/pull/21286
- [x] questdb https://github.com/Homebrew/homebrew-core/pull/24037
- [x] ringojs https://github.com/Homebrew/homebrew-core/pull/19919
- [x] scala@2.10 https://github.com/Homebrew/homebrew-core/pull/24041
- [x] scala@2.11 https://github.com/Homebrew/homebrew-core/pull/24045
- [x] scm-manager https://github.com/Homebrew/homebrew-core/pull/20510
- [x] tomee-webprofile https://github.com/Homebrew/homebrew-core/pull/24038

Broken at build time (`brew install --build-from-source` fails):
- [x] abcl https://github.com/Homebrew/homebrew-core/pull/28312
- [x] avian https://github.com/Homebrew/homebrew-core/pull/28313
- [x] aws-apigateway-importer https://github.com/Homebrew/homebrew-core/pull/28315
- [x] drip https://github.com/Homebrew/homebrew-core/pull/28314
- [x] gssh https://github.com/Homebrew/homebrew-core/pull/28316
- [x] henplus https://github.com/Homebrew/homebrew-core/pull/28317
- [x] hexgui https://github.com/Homebrew/homebrew-core/pull/28318
- [x] lcm https://github.com/Homebrew/homebrew-core/pull/28319
- [x] mysql-cluster https://github.com/Homebrew/homebrew-core/pull/28320
- [ ] opentsdb https://gist.github.com/ilovezfs/5b4c6c9a21c4eb10386606ee439549ad
- [x] pillar https://github.com/Homebrew/homebrew-core/pull/28321
- [x] voldemort https://github.com/Homebrew/homebrew-core/pull/28322
- [x] wartremover https://github.com/Homebrew/homebrew-core/pull/28323
- [x] duck https://github.com/Homebrew/homebrew-core/pull/19938
- [x] gosu https://github.com/Homebrew/homebrew-core/pull/21414
- [x] groovyserv https://github.com/Homebrew/homebrew-core/pull/25189
- [x] htmlcleaner https://github.com/Homebrew/homebrew-core/pull/20785
- [x] libbluray https://github.com/Homebrew/homebrew-core/pull/21646
- [x] swagger-codegen https://github.com/Homebrew/homebrew-core/pull/22052

Works at run time (`brew test` passes if it has a test block) and build time status is unknown or not applicable (bottle unneeded) (NB: may still be broken if the formula has no test block):

- [x] activemq
- [x] akka
- [x] allure
- [x] apache-archiva
- [x] apache-ctakes
- [x] apache-geode
- [x] antlr
- [x] artifactory
- [x] asciidoctorj
- [x] atlassian-cli
- [x] auto-scaling
- [x] avro-tools
- [x] aws-cfn-tools
- [x] aws-cloudsearch
- [x] aws-elasticache
- [x] aws-sns-cli
- [x] cfr-decompiler
- [x] checkstyle
- [x] civl
- [x] clojure
- [x] clojurescript
- [x] closure-compiler
- [x] closure-stylesheets
- [x] cloud-watch
- [x] commandbox (https://github.com/Homebrew/homebrew-core/pull/20661)
- [x] crash
- [x] cromwell
- [x] crowdin
- [x] datomic
- [x] davmail
- [x] denominator
- [x] dependency-check
- [x] derby
- [x] dex2jar
- [x] digdag
- [x] ec2-ami-tools
- [x] ec2-api-tools
- [x] elasticsearch
- [x] elb-tools
- [x] excel-compare
- [x] findbugs
- [x] flume
- [x] fmpp
- [x] fop
- [x] frege-repl
- [x] frege
- [x] galen
- [x] gcviewer
- [x] geoserver
- [x] git-credential-manager
- [x] gitbucket
- [x] glassfish
- [x] golo
- [x] gradle
- [x] grails@2.5
- [x] grakn
- [x] groovy
- [x] groovysdk
- [x] gwt
- [x] h2
- [x] hadoop
- [x] hive
- [x] i2p
- [x] ivy
- [x] jadx
- [x] jasmin
- [x] jbake
- [x] jboss-as
- [x] jboss-forge
- [x] jenkins-lts
- [x] jenkins
- [x] jenv
- [x] jetty-runner
- [x] jetty
- [x] jflex
- [x] jruby
- [x] js-test-driver
- [x] jsawk
- [x] jslint4java
- [x] jsonschema2pojo
- [x] jvmtop
- [x] kawa
- [x] kotlin
- [x] languagetool
- [x] launch4j
- [x] linkerd
- [x] logstash@2.4
- [x] luciddb
- [x] mahout
- [x] mockserver
- [x] narwhal
- [x] nifi
- [x] orc-tools
- [x] orientdb
- [x] pax-runner
- [x] pig
- [x] plantuml
- [x] predictionio
- [x] presto
- [x] procyon-decompiler
- [x] proguard
- [x] rds-command-line-tools
- [x] redpen
- [x] rhino
- [x] sbt
- [x] sbt@0.13
- [x] sbuild
- [x] scala
- [x] scalastyle
- [x] sdedit
- [x] selenium-server-standalone
- [x] selenium-server-standalone@2.45
- [x] sisc-scheme
- [x] sjk
- [x] skinny
- [x] smali
- [x] solr
- [x] solr@5.5
- [x] sonarlint
- [x] sonarqube
- [x] spring-loaded
- [x] spring-roo
- [x] sqoop
- [x] stanford-ner
- [x] stanford-parser
- [x] tailor
- [x] tee-clc
- [x] tika
- [x] titan-server
- [x] tomcat
- [x] tomcat@6
- [x] tomcat@7
- [x] typesafe-activator
- [x] umlet
- [x] vault-cli
- [x] vert.x
- [x] vnu
- [x] walkmod
- [x] wildfly-as
- [x] xmlsectool
- [x] yarn
- [x] yuicompressor
- [x] zanata-client

Non-bottle-unneeded formulae with non-optional `depends_on :java` not yet using a hard Java 8 req:

- [x] ant
- [x] antlr@2
- [x] apache-forrest
- [x] apollo
- [x] bdsup2sub
- [x] beast
- [x] couchdb-lucene
- [x] duck
- [x] fits
- [x] giter8
- [x] graphviz
- [x] hbase
- [x] jdnssec-tools
- [x] jhipster
- [x] jpcsp
- [x] jsvc
- [x] libphonenumber
- [x] libreadline-java
- [x] maven@3.1
- [x] maven@3.2
- [x] mesos
- [x] nesc
- [x] pmd 
- [x] shogun
- [x] sleuthkit
- [x] tomcat-native
- [x] tomcat@8
- [x] voms

**Comment by ilovezfs** [comment]
I've added the status of all bottle unneeded ones above. The bottle-needed ones still need to be assessed.

**Comment by SethTisue** [comment]
`scala` is already reasonably JDK9 friendly, as of last week's Scala 2.12.4 release.

`scala@2.10` will be reasonably JDK9 friendly once Scala 2.10.7 becomes available (within a few weeks, we think). `scala@2.11`, ditto for Scala 2.11.12.

`sbt` will be reasonably JDK9 friendly in its upcoming 1.0.3 release and, in the `sbt@0.13` line, in its 0.13.17 release. Neither release exists yet.

**Comment by ilovezfs** [comment]
Thanks @SethTisue! That's very helpful.

**Comment by botic** [comment]
The earliest version of ringojs with JDK 9 support is likely 2.x (~ Q1/Q2 2018) since it depends on full support of Java 9 in Jetty and Rhino. So testing RingoJS with Java 8 would solve the issue. The broken test: https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/11720/

**Comment by jrwiegand** [comment]
Works great, thanks for all the work you did on this.

**Comment by Moisan** [comment]
`javarepl` issue is [known upstream](https://github.com/albertlatacz/java-repl/issues/125)

**Comment by yetisyny** [comment]
`libbluray` should be listed, the current stable version 1.0.2 does not yet work with java 9 and homebrew is set to only let it work with java 8, although they are working on making the next stable version of `libbluray` compatible with java 9 and if you get a HEAD build it probably would work already (almost all of the recent commits have been related to java 9 compatibility but I am not sure if it is 100% compatible yet). You can try to see if it is compatible with the --HEAD option with `libbluray` in homebrew.

Anyway I would add this to your list which does not list `libbluray`. I am very sure that when the next version of `libbluray` comes out it will be java 9 compatible, after looking at their recent commit history and their history of releases as well as their mailing list. It should be added to the first list, with a checkmark next to it for being fixed, since it is working right now by being pinned to java 8. This way you can keep track of it along with all the other java-dependent formulae with issues with java 9.

**Comment by ilovezfs** [comment]
@yetisyny thanks. added.

**Comment by yetisyny** [comment]
Java is moving to rapid-release, with major releases every 6 months and long-term support releases every 3 years, a major change from the past. Java SE 8 will stop getting security updates in January 2019. Java SE 9 will stop getting updates even sooner, next month, in March 2018, when Java SE 10 (18.3) comes out and supersedes it, and again, in September 2018, Java SE 10 (18.3) will stop getting updates and Java SE 11 (18.9-LTS) will come out and supersede it. Java SE 11 (18.9-LTS) will be supported for 3 years most likely, probably until September 2021, so really we will want everything working on Java 11 since it will be out a lot longer than Java 9 or 10 are going to last.

For more information see [http://www.oracle.com/technetwork/java/javase/eol-135779.html](http://www.oracle.com/technetwork/java/javase/eol-135779.html). Anyway my point here is, in just one month from now, Java 9 will be history and Java 10 will be out. And also Java 8, the version that pretty much everything is compatible with, is going to stop getting updates in January next year and then everyone will be forced to either update to one of these rolling-release versions that comes out every 6 months or a long-term support release like Java 11 which will be supported for 3 years.

So in other words, it’s a much faster release cycle starting next month and everybody will have to get things compatible with new Java versions every 6 months. Which means a lot more work for people keeping formulae that use Java compatible with the latest Java versions, so please get ready for it.

**Comment by ilovezfs** [comment]
We'll remove whatever doesn't work. There's nothing to get ready for.

**Comment by ilovezfs** [comment]
We'll remove whatever doesn't work. There's nothing to get ready for.

**Comment by apjanke** [comment]
Well, darn. Here's `apache-flink`:

```
  depends_on :java => "1.7+"
```

We don't have a way of expressing "at least Java 1.7, but no greater than 1.8".

Part of me wants to say "just use '1.8'", because Java 1.7 is pretty old and obsolete at this point. (It went EOL for non-enterprise users way back in April 2015.) But that's not the part of me that's running older machines.

**Comment by ilovezfs** [comment]
@apjanke grep through core. These have all been handled so far with a hard Java 8 dep.

**Comment by ilovezfs** [comment]
Note that in the case of apache-flink that has proven insufficient, so it will soon be removed if it cannot coexist peacefully on a dual Java 8 / Java 9 system. See
https://github.com/Homebrew/homebrew-core/pull/21823
https://github.com/Homebrew/homebrew-core/pull/21970
https://github.com/Homebrew/homebrew-core/pull/21596

**Comment by apjanke** [comment]
Sounds good.

I agree with a hard Java 8 dep. IMHO, Homebrew should not be supporting users running Java 1.7, since it is no longer receiving security updates.

**Comment by ilovezfs** [comment]
@apjanke FYI I added a new section above with the catchy title

>Non-bottle-unneeded formulae with non-optional depends_on :java not yet using a hard Java 8 req and of unknown build status with Java 9:

**Comment by apjanke** [comment]
Ooh, more testing to do!

**Comment by yetisyny** [comment]
Java SE 10 has been out for almost a week, since March 20th, and Java SE 9 is no longer supported AT ALL, with all Java SE 9 users being automatically updated to Java SE 10, at least for the JRE (maybe not with the JDK, I think JDK still requires manual updates).

Java SE 8 continues to have public security updates, so basically the 2 versions of Java that are out now that it is important to check for compatibility with are 8 and 10. See [http://www.oracle.com/technetwork/java/eol-135779.html](http://www.oracle.com/technetwork/java/eol-135779.html).

I suggest that it’s probably time to close this tracking bug and open a new one for Java SE 10.

**Comment by ilovezfs** [comment]
@yetisyny we know, but thanks for the heads up.

**Comment by Jahhein** [comment]
@yetisyny In a few more weeks it will no longer be called Java. [Java EE rebranded to Jakarta EE](https://mmilinkov.wordpress.com/2018/02/26/and-the-name-is/)

The deal with oracle is the same with Java SE. Don't call it Java anymore.

**Comment by apjanke** [comment]
`apache-flink` is still broken under `apache-flink` 1.5.0, but it looks like a simple fix in its startup wrapper scripts. They're just using a raw `java` call instead of using `$JAVA_HOME` in one place. Upstream bug report: https://issues.apache.org/jira/browse/FLINK-9445

**Comment by apjanke** [comment]
It looks like `embulk` is fixed now, by 4219b20e5b14c653deafe6b75395d92138adb3a6.

**Comment by apjanke** [comment]
Following up on `ecj` and #24029: What do you all think about just dropping `ecj` and its Java support in the earlier GCCs (`gcc@5`, `gcc@6`, `gcc@4.9`)? Vendoring it in wouldn't be very hard, but given the low usage (<1% of users install `ecj`, per https://github.com/Homebrew/homebrew-core/pull/24029#issuecomment-365229689) it doesn't seem worth the maintenance effort to me. I'd be kinda surprised if any significant number of people were actually using the old GCC Java support, instead of just picking it up with an optimistic `--with-all-languages`. I'm in favor of dropping it, and will put in a PR to do so if you're amenable.

**Comment by apjanke** [comment]
Are you sure `fastbit` is due to a JDK 9 issue? The error in your Gist doesn't look Java-related to me:

```
checking if compiler supports __sync_add_and_fetch for 64-bit integers... yes
checking if the default FlexLexer.h defines yywrap... configure: error: no, must have FlexLexer.h to continue
```

... and `brew install -s fastbit` works fine on my macOS 10.11 machine with JDK 9 and 10 installed.

**Comment by apjanke** [comment]
`joshua`'s URL has been fixed; `brew install -s joshua` works for me with JDK 9 & 10 installed; it can be removed from this list.

**Comment by apjanke** [comment]
Fixes incoming:

abcl: #28312
avian: #28313
drip: #28314 
aws-apigateway-importer: #28315
gssh: #28316
henplus: #28317
hexgui: #28318
lcm: #28319
mysql-cluster: #28320 

These all address the issue by pinning it to Java 8 specifically, either at build or run time or both. I think that's really the only way to go at this point, until upstreams all make the leap to Java 9/10/11.

**Comment by apjanke** [comment]
jooby-bootstrap: #28311

**Comment by apjanke** [comment]
`upscaledb` works for me now, on a 10.13 system with Java 8 & 9 & 10.

Looks to me like the build error here was not Java related, and it can probably be removed from the list:

```
In file included from 3changeset/changeset.cc:18:
./2worker/worker.h:83:16: error: use of class template 'boost::asio::strand' requires template arguments
  boost::asio::strand strand;
               ^
/usr/local/include/boost/asio/strand.hpp:29:7: note: template is declared here
class strand
      ^
1 error generated.
make[2]: *** [3changeset/changeset.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [install-recursive] Error 1
make: *** [install-recursive] Error 1
```

**Comment by apjanke** [comment]
Fixes incoming:

pillar: #28321
voldemort: #28322
wartremover: #28323

That's all the low-hanging fruit.

**Comment by ilovezfs** [comment]
I think fastbit was fixed in https://github.com/Homebrew/homebrew-core/pull/25228 and unrelated to Java.

Thanks for the all the awesome work @apjanke!

So I think we have outstanding problems with apache-flink, ecj, and opentsdb?

**Comment by apjanke** [comment]
You're welcome!

hbase needs to go back on the "to do" list, too. Its own install and test succeed. But it emits warnings indicating it's not really ready for Java 9.

```

$ start-hbase.sh
Java HotSpot(TM) 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
starting master, logging to /usr/local/var/log/hbase/hbase-janke-master-angharad.local.out
Java HotSpot(TM) 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Java HotSpot(TM) 64-Bit Server VM warning: Ignoring option PermSize; support was removed in 8.0
Java HotSpot(TM) 64-Bit Server VM warning: Ignoring option MaxPermSize; support was removed in 8.0
```

And when actually exercising it, as is done in the opentsdb post_install, it throws Java 9 related errors.

```
$ brew install -s opentsdb
==> Downloading https://github.com/OpenTSDB/opentsdb/releases/download/v2.3.0/opentsdb-2.3.0.tar.gz
Already downloaded: /Users/janke/Library/Caches/Homebrew/opentsdb-2.3.0.tar.gz
==> ./configure --disable-silent-rules --prefix=/usr/local/Cellar/opentsdb/2.3.0_1 --mandir=/usr/local/Cellar/opentsdb/2.
==> make
==> make install
==> /usr/local/opt/hbase/bin/start-hbase.sh
==> /usr/local/Cellar/opentsdb/2.3.0_1/share/opentsdb/tools/create_table_with_env.sh
Last 15 lines from /Users/janke/Library/Logs/Homebrew/opentsdb/post_install.02.create_table_with_env.sh:

Java HotSpot(TM) 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.jruby.java.invokers.RubyToJavaInvoker (file:/usr/local/Cellar/hbase/1.2.6/libexec/lib/jruby-complete-1.6.8.jar) to method java.lang.Object.registerNatives()
WARNING: Please consider reporting this to the maintainers of org.jruby.java.invokers.RubyToJavaInvoker
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
ArgumentError: wrong number of arguments (0 for 1)
  method_added at file:/usr/local/Cellar/hbase/1.2.6/libexec/lib/jruby-complete-1.6.8.jar!/builtin/javasupport/core_ext/object.rb:10
  method_added at file:/usr/local/Cellar/hbase/1.2.6/libexec/lib/jruby-complete-1.6.8.jar!/builtin/javasupport/core_ext/object.rb:129
       Pattern at file:/usr/local/Cellar/hbase/1.2.6/libexec/lib/jruby-complete-1.6.8.jar!/builtin/java/java.util.regex.rb:2
        (root) at file:/usr/local/Cellar/hbase/1.2.6/libexec/lib/jruby-complete-1.6.8.jar!/builtin/java/java.util.regex.rb:1
       require at org/jruby/RubyKernel.java:1062
        (root) at file:/usr/local/Cellar/hbase/1.2.6/libexec/lib/jruby-complete-1.6.8.jar!/builtin/java/java.util.regex.rb:42
        (root) at /usr/local/opt/hbase/libexec/bin/../bin/hirb.rb:38
==> /usr/local/opt/hbase/bin/stop-hbase.sh
```

Note that those are coming from the hbase Java code, not opentsdb itself.

**Comment by ilovezfs** [comment]
ecj deleted!

I'm closing this out since we have open PRs for all the outstanding problems. Thanks for the help everyone, and especially @apjanke, the MVP!

**Comment by apjanke** [comment]
🎉
